lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt

```diff
@@ -0,0 +1,288 @@
+            }
+            // object feed into inputObject (i.e. hydrating with a $source object)
+            else if (sourceType == ObjectField && unwrappedHydrationSourceFieldType is GraphQLObjectType && unwrappedActorFieldArgType is GraphQLInputObjectType) {
+                return validateInputObjectArg(unwrappedHydrationSourceFieldType, unwrappedActorFieldArgType, parent, overallField, remoteArg, hydration, actorFieldName)
+            }
+            // inputObject feed into inputObject (i.e. hydrating with an $argument object)
+            else if (sourceType == FieldArgument && unwrappedHydrationSourceFieldType is GraphQLInputObjectType && unwrappedActorFieldArgType is GraphQLInputObjectType) {
+                if (unwrappedHydrationSourceFieldType.name != unwrappedActorFieldArgType.name) {
+                    return NadelSchemaValidationError.IncompatibleHydrationArgumentType(
+                            parent,
+                            overallField,
+                            remoteArg,
+                            hydrationSourceFieldType,
+                            actorFieldArgType,
+                            actorFieldName
+                    )
+                }
+                return null
+            }
+            // if enums they must be the same
+            else if (unwrappedHydrationSourceFieldType is GraphQLEnumType && unwrappedActorFieldArgType is GraphQLEnumType){
+                if (unwrappedHydrationSourceFieldType.name != unwrappedActorFieldArgType.name) {
+                    return NadelSchemaValidationError.IncompatibleHydrationArgumentType(
+                            parent,
+                            overallField,
+                            remoteArg,
+                            hydrationSourceFieldType,
+                            actorFieldArgType,
+                            actorFieldName
+                    )
+                }
+                return null
+            }
+            // Any other types are not assignable
+            return NadelSchemaValidationError.IncompatibleHydrationArgumentType(
+                    parent,
+                    overallField,
+                    remoteArg,
+                    hydrationSourceFieldType,
+                    actorFieldArgType,
+                    actorFieldName
+            )
+        }
+
+        private fun validateScalarArg(
+                hydrationSourceFieldType: GraphQLType,
+                actorFieldArgType: GraphQLType,
+                parent: NadelServiceSchemaElement,
+                overallField: GraphQLFieldDefinition,
+                remoteArg: RemoteArgumentDefinition,
+                actorFieldName: String
+        ): NadelSchemaValidationError? {
+            if (hydrationSourceFieldType.unwrapNonNull() is GraphQLScalarType && actorFieldArgType.unwrapNonNull() is GraphQLScalarType) {
```

Comment:
Please add `hydrationSourceFieldType.unwrapNonNull()` as variables so you don't have to manually cast it below.